src/compiler.js

```diff
@@ -0,0 +1,91 @@
+import { normalize, join } from 'path'
+import { promisify, Promise } from 'bluebird'
+import { normalizeOptions } from './options'
+import { readFile, writeFile, createReadStream } from 'fs'
+import { spawn } from 'child_process'
+import * as logger from './logger'
+import { dequote } from './util'
+
+const isWindows = process.platform === 'win32'
+const isBsd = Boolean(~process.platform.indexOf('bsd'))
+const make = isWindows ? 'vcbuild.bat' : isBsd ? 'gmake' : 'make'
+const configure = isWindows ? 'configure' : './configure'
+const readFileAsync = promisify(readFile)
+const writeFileAsync = promisify(writeFile)
+
+export class NexeCompiler {
+  constructor (options) {
+    options = this.options = normalizeOptions(options)
+    logger.setLevel(options.loglevel)
+    this.log = logger
+    this.python = options.python
+    this.configure = options.configure
+    this.make = isWindows ? options.vcBuild : options.make
+    this.src = join(options.temp, options.version)
+    this.env = Object.assign({}, process.env)
+    this.files = []
+
+    this.readFileAsync = async (file) => {
+      const cachedFile = this.files.find(x => normalize(x.filename) === normalize(file))
+      if (cachedFile) {
+        return Promise.resolve(cachedFile)
+      }
+      this.files.push({
+        filename: file,
+        contents: await readFileAsync(join(this.src, file), 'utf-8').catch(e => {
+          if (e.code !== 'ENOENT') {
+            throw e
+          }
+          return ''
+        })
+      })
+      return this.readFileAsync(file)
```

Comment:
Rather than recurse, I would flip the process in this:

```js
let cachedFile = this.files.find(x => normalize(x.filename) === normalize(file))
if (!cachedFile) { 
  cachedFile = { filename: file, contents: ... };
  this.files.push(cachedFile);
}
return cachedFile;
```